TestProcesses failing on CI for darwin

The `TestProcesses` is constantly [failing on CI](https://beats-ci.elastic.co/blue/organizations/jenkins/Library%2Fgo-sysinfo-mbp/detail/PR-144/34/pipeline/152) at least for https://github.com/elastic/go-sysinfo/pull/144, but it seems to be happening on other PRs as well, even though it passes locally.

I don´t have a mac to investigate it further, I just got someone else to try it and as I said, itś working locally.


```
[2023-01-23T13:49:10.877Z] === RUN   TestProcesses

[2023-01-23T13:49:10.877Z]     process_darwin_test.go:115: empty exec

[2023-01-23T13:49:10.877Z] --- FAIL: TestProcesses (0.00s)

[2023-01-23T13:49:10.877Z] FAIL

[2023-01-23T13:49:10.877Z] FAIL	github.com/elastic/go-sysinfo/providers/darwin	0.671s
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -19,6 +19,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Fixed
 
+- On darwin without CGO `process.Info()` could fail, but would not return the error. [#150](https://github.com/elastic/go-sysinfo/pull/150)
+
 ## [1.9.0]
 
 ### Added

--- a/providers/darwin/process_darwin.go
+++ b/providers/darwin/process_darwin.go
@@ -23,9 +23,11 @@ package darwin
 import (
 	"bytes"
 	"encoding/binary"
+	"errors"
 	"fmt"
 	"os"
 	"strconv"
+	"syscall"
 	"time"
 
 	"golang.org/x/sys/unix"
@@ -178,7 +180,12 @@ var nullTerminator = []byte{0}
 func kern_procargs(pid int, p *process) error {
 	data, err := unix.SysctlRaw("kern.procargs2", pid)
 	if err != nil {
-		return nil
+		if errors.Is(err, syscall.EINVAL) {
+			// sysctl returns "invalid argument" for both "no such process"
+			// and "operation not permitted" errors.
+			return fmt.Errorf("no such process or operation not permitted: %w", err)
+		}
+		return err
 	}
 	buf := bytes.NewBuffer(data)
 

--- a/providers/darwin/process_darwin_test.go
+++ b/providers/darwin/process_darwin_test.go
@@ -21,8 +21,10 @@
 package darwin
 
 import (
+	"errors"
 	"os"
 	"os/exec"
+	"syscall"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
@@ -97,33 +99,44 @@ func TestProcessEnvironmentInternal(t *testing.T) {
 
 func TestProcesses(t *testing.T) {
 	var s darwinSystem
-	ps, err := s.Processes()
+	processes, err := s.Processes()
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	pinfo, err := ps[0].Info()
-	if err != nil {
-		t.Fatal(err)
+	var count int
+	for _, proc := range processes {
+		processInfo, err := proc.Info()
+		switch {
+		// Ignore processes that no longer exist or that cannot be accessed.
+		case errors.Is(err, syscall.ESRCH),
+			errors.Is(err, syscall.EPERM),
+			errors.Is(err, syscall.EINVAL):
+			continue
+		case err != nil:
+			t.Fatalf("failed to get process info for PID=%d: %v", proc.PID(), err)
+		default:
+			count++
+		}
+
+		if processInfo.PID == 0 {
+			t.Fatalf("empty pid in %#v", processInfo)
+		}
+
+		if processInfo.Exe == "" {
+			t.Fatalf("empty exec in %#v", processInfo)
+		}
+
+		u, err := proc.User()
+		require.NoError(t, err)
+
+		require.NotEmpty(t, u.UID)
+		require.NotEmpty(t, u.EUID)
+		require.NotEmpty(t, u.SUID)
+		require.NotEmpty(t, u.GID)
+		require.NotEmpty(t, u.EGID)
+		require.NotEmpty(t, u.SGID)
 	}
 
-	if pinfo.PID == 0 {
-		t.Fatal("empty pid")
-	}
-
-	if pinfo.Exe == "" {
-		t.Fatal("empty exec")
-	}
-
-	u, err := ps[0].User()
-	require.NoError(t, err)
-
-	require.NotEmpty(t, u.UID)
-	require.NotEmpty(t, u.EUID)
-	require.NotEmpty(t, u.SUID)
-	require.NotEmpty(t, u.GID)
-	require.NotEmpty(t, u.EGID)
-	require.NotEmpty(t, u.SGID)
-
-	t.Log(ps[0].Info())
+	assert.NotZero(t, count, "failed to get process info for any processes")
 }

--- a/system_test.go
+++ b/system_test.go
@@ -292,14 +292,17 @@ func TestProcesses(t *testing.T) {
 	t.Log("Found", len(procs), "processes.")
 	for _, proc := range procs {
 		info, err := proc.Info()
-		if err != nil {
-			if errors.Is(err, fs.ErrPermission) || errors.Is(err, syscall.ESRCH) {
-				// The process may no longer exist by the time we try fetching
-				// additional information so ignore ESRCH (no such process).
-				continue
-			}
-			t.Fatal(err)
+		switch {
+		// Ignore processes that no longer exist or that cannot be accessed.
+		case errors.Is(err, syscall.ESRCH),
+			errors.Is(err, syscall.EPERM),
+			errors.Is(err, syscall.EINVAL),
+			errors.Is(err, fs.ErrPermission):
+			continue
+		case err != nil:
+			t.Fatalf("failed to get process info for PID=%d: %v", proc.PID(), err)
 		}
+
 		t.Logf("pid=%v name='%s' exe='%s' args=%+v ppid=%d cwd='%s' start_time=%v",
 			info.PID, info.Name, info.Exe, info.Args, info.PPID, info.CWD,
 			info.StartTime)

